src/Bandits/BanditEvaluator.php

```diff
@@ -0,0 +1,251 @@
+    {
+        $numberOfActions = count($actionScores);
+        $bestActionKey = array_keys($actionScores, max($actionScores))[0];
+
+
+        $minProbability = $probabilityFloor / $numberOfActions;
+
+        $weights = array_filter(
+            $actionScores,
+            function ($key) use ($bestActionKey) {
+                return $key !== $bestActionKey;
+            },
+            ARRAY_FILTER_USE_KEY
+        );
+
+        $bestScore = $actionScores[$bestActionKey];
+        $weights = array_map(
+            function ($score) use ($minProbability, $bestScore, $gamma, $numberOfActions) {
+                return max($minProbability, 1.0 / ($numberOfActions + $gamma * ($bestScore - $score)));
+            },
+            $weights
+        );
+
+        $remainingWeight = max(0.0, 1.0 - array_sum($weights));
+        $weights[$bestActionKey] = $remainingWeight;
+
+        return $weights;
+    }
+
+    /**
+     * @param string $flagKey
+     * @param string $subjectKey
+     * @param array<string, float> $actionWeights
+     * @return string
+     * @throws BanditEvaluationException
+     */
+    private function selectAction(string $flagKey, string $subjectKey, array $actionWeights): string
+    {
+        // Use a list of key-value pairs to make sorting easier.
+        $weightPairs = array_map(
+            fn($key) => new ActionValue($key, $actionWeights[$key]),
+            array_keys($actionWeights)
+        );
+
+        // usort sorts in place.
+        usort(
+            $weightPairs,
+            function (ActionValue $a, ActionValue $b) use ($subjectKey, $flagKey) {
+                $aValue = Sharder::getShard("$flagKey-$subjectKey-{$a->action}", $this->totalShards);
+                $bValue = Sharder::getShard("$flagKey-$subjectKey-{$b->action}", $this->totalShards);
+
+                return $aValue === $bValue ? 0 : (($aValue < $bValue) ? -1 : 1);
+            }
+        );
+
+        // Bucket the user
+        $shard = Sharder::getShard("$flagKey-$subjectKey", $this->totalShards);
+        $cumulativeWeight = 0.0;
+        $shardValue = $shard / $this->totalShards;
+
+        foreach ($weightPairs as $weightData) {
+            $cumulativeWeight += $weightData->value;
+            if ($cumulativeWeight > $shardValue) {
+                return $weightData->action;
+            }
+        }
+
+        throw new BanditEvaluationException(
+            "[Eppo SDK] No action selected for $flagKey $subjectKey"
+        );
+    }
+
+    /**
+     * @param array<NumericAttributeCoefficient> $coefficients
+     * @param array<string, float> $attributes
+     * @return float
+     */
+    public static function scoreNumericAttributes(array $coefficients, array $attributes): float
+    {
+        $score = 0.0;
+        foreach ($coefficients as $coefficient) {
+            $attributeKey = $coefficient->attributeKey;
+            if (array_key_exists($attributeKey, $attributes)) {
```

Comment:
I think we also want to make sure that the value is numeric, in case `NULL` or a non-number is inadvertently passed in as a numeric attribute upstream. If it is not, we should ignore it and use missing value coefficient (vs throwing an error).

Reply:
good catch.
Added a test that causes code to throw before fix.
![image](https://github.com/Eppo-exp/php-sdk/assets/9689485/d2536263-f372-4232-9ef1-019c34bdf47b)